File: leetcode/c++/0053-maximum-subarray.cc
// https://leetcode.com/problems/maximum-subarray/
// Easy
#include<vector>

using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxSoFar = nums[0], size = nums.size(), curr = max(nums[0], 0);
        for (int i = 1; i < size; ++i) {
            maxSoFar = max(maxSoFar, curr + nums[i]);
            curr = max(curr+nums[i], 0);
        }
        return maxSoFar;
    }
};

File: leetcode/python/0053-maximum-subarray.py
# https://leetcode.com/problems/maximum-subarray/
# Easy

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maxSoFar, currSum = nums[0], 0

        for n in nums:
            currSum = max(n, currSum + n)
            maxSoFar = max(maxSoFar, currSum)
                
        return maxSoFar
            